mesonbuild/cmake/generator.py

```diff
@@ -130,6 +131,10 @@ def target_file(arg: str) -> str:
         'TARGET_NAME_IF_EXISTS': lambda x: x if x in trace.targets else '',
         'TARGET_PROPERTY': target_property,
         'TARGET_FILE': target_file,
+
+        # Configurations (Debug, Release...)
+        # CMake is case-insensitive
+        'CONFIG': lambda x: '1' if x.upper() == trace.env.coredata.get_option(OptionKey('buildtype')).upper() else '0'
```

Comment:
The CONFIG value from cmake will be CMAKE_BUILD_TYPE, correct?

Reply:
In `CMake` yes. But in `meson` it is the `buildtype`. The `meson` `cmake` module does not set `CMAKE_BUILD_TYPE`.

Reply:
Here is what `conan` will produce for `ZLIB` (follow the includes):

<details>
<summary>ZLIBConfig.cmake</summary>

```
# Requires CMake > 3.15
if(${CMAKE_VERSION} VERSION_LESS "3.15")
    message(FATAL_ERROR "The 'CMakeDeps' generator only works with CMake >= 3.15")
endif()

if(ZLIB_FIND_QUIETLY)
    set(ZLIB_MESSAGE_MODE VERBOSE)
else()
    set(ZLIB_MESSAGE_MODE STATUS)
endif()

include(${CMAKE_CURRENT_LIST_DIR}/cmakedeps_macros.cmake)
include(${CMAKE_CURRENT_LIST_DIR}/ZLIBTargets.cmake)
include(CMakeFindDependencyMacro)

check_build_type_defined()

foreach(_DEPENDENCY ${zlib_FIND_DEPENDENCY_NAMES} )
    # Check that we have not already called a find_package with the transitive dependency
    if(NOT ${_DEPENDENCY}_FOUND)
        find_dependency(${_DEPENDENCY} REQUIRED ${${_DEPENDENCY}_FIND_MODE})
    endif()
endforeach()

set(ZLIB_VERSION_STRING "1.2.13")
set(ZLIB_INCLUDE_DIRS ${zlib_INCLUDE_DIRS_RELEASE} )
set(ZLIB_INCLUDE_DIR ${zlib_INCLUDE_DIRS_RELEASE} )
set(ZLIB_LIBRARIES ${zlib_LIBRARIES_RELEASE} )
set(ZLIB_DEFINITIONS ${zlib_DEFINITIONS_RELEASE} )

# Only the first installed configuration is included to avoid the collision
foreach(_BUILD_MODULE ${zlib_BUILD_MODULES_PATHS_RELEASE} )
    message(${ZLIB_MESSAGE_MODE} "Conan: Including build module from '${_BUILD_MODULE}'")
    include(${_BUILD_MODULE})
endforeach()
```

</details>




<details>
<summary>ZLIBTargets.cmake</summary>

```
# Load the debug and release variables
file(GLOB DATA_FILES "${CMAKE_CURRENT_LIST_DIR}/ZLIB-*-data.cmake")

foreach(f ${DATA_FILES})
    include(${f})
endforeach()

# Create the targets for all the components
foreach(_COMPONENT ${zlib_COMPONENT_NAMES} )
    if(NOT TARGET ${_COMPONENT})
        add_library(${_COMPONENT} INTERFACE IMPORTED)
        message(${ZLIB_MESSAGE_MODE} "Conan: Component target declared '${_COMPONENT}'")
    endif()
endforeach()

if(NOT TARGET ZLIB::ZLIB)
    add_library(ZLIB::ZLIB INTERFACE IMPORTED)
    message(${ZLIB_MESSAGE_MODE} "Conan: Target declared 'ZLIB::ZLIB'")
endif()
# Load the debug and release library finders
file(GLOB CONFIG_FILES "${CMAKE_CURRENT_LIST_DIR}/ZLIB-Target-*.cmake")

foreach(f ${CONFIG_FILES})
    include(${f})
endforeach()
```

</details>




<details>
<summary>ZLIB-Target-release.cmake</summary>

```
# Avoid multiple calls to find_package to append duplicated properties to the targets
include_guard()########### VARIABLES #######################################################################
#############################################################################################
set(zlib_FRAMEWORKS_FOUND_RELEASE "") # Will be filled later
conan_find_apple_frameworks(zlib_FRAMEWORKS_FOUND_RELEASE "${zlib_FRAMEWORKS_RELEASE}" "${zlib_FRAMEWORK_DIRS_RELEASE}")

set(zlib_LIBRARIES_TARGETS "") # Will be filled later


######## Create an interface target to contain all the dependencies (frameworks, system and conan deps)
if(NOT TARGET zlib_DEPS_TARGET)
    add_library(zlib_DEPS_TARGET INTERFACE IMPORTED)
endif()

set_property(TARGET zlib_DEPS_TARGET
             APPEND PROPERTY INTERFACE_LINK_LIBRARIES
             $<$<CONFIG:Release>:${zlib_FRAMEWORKS_FOUND_RELEASE}>
             $<$<CONFIG:Release>:${zlib_SYSTEM_LIBS_RELEASE}>
             $<$<CONFIG:Release>:>)

####### Find the libraries declared in cpp_info.libs, create an IMPORTED target for each one and link the
####### zlib_DEPS_TARGET to all of them
conan_package_library_targets("${zlib_LIBS_RELEASE}"    # libraries
                              "${zlib_LIB_DIRS_RELEASE}" # package_libdir
                              "${zlib_BIN_DIRS_RELEASE}" # package_bindir
                              "${zlib_LIBRARY_TYPE_RELEASE}"
                              "${zlib_IS_HOST_WINDOWS_RELEASE}"
                              zlib_DEPS_TARGET
                              zlib_LIBRARIES_TARGETS  # out_libraries_targets
                              "_RELEASE"
                              "zlib"    # package_name
                              "${zlib_NO_SONAME_MODE_RELEASE}")  # soname

# FIXME: What is the result of this for multi-config? All configs adding themselves to path?
set(CMAKE_MODULE_PATH ${zlib_BUILD_DIRS_RELEASE} ${CMAKE_MODULE_PATH})

########## GLOBAL TARGET PROPERTIES Release ########################################
    set_property(TARGET ZLIB::ZLIB
                 APPEND PROPERTY INTERFACE_LINK_LIBRARIES
                 $<$<CONFIG:Release>:${zlib_OBJECTS_RELEASE}>
                 $<$<CONFIG:Release>:${zlib_LIBRARIES_TARGETS}>
                 )

    if("${zlib_LIBS_RELEASE}" STREQUAL "")
        # If the package is not declaring any "cpp_info.libs" the package deps, system libs,
        # frameworks etc are not linked to the imported targets and we need to do it to the
        # global target
        set_property(TARGET ZLIB::ZLIB
                     APPEND PROPERTY INTERFACE_LINK_LIBRARIES
                     zlib_DEPS_TARGET)
    endif()

    set_property(TARGET ZLIB::ZLIB
                 APPEND PROPERTY INTERFACE_LINK_OPTIONS
                 $<$<CONFIG:Release>:${zlib_LINKER_FLAGS_RELEASE}>)
    set_property(TARGET ZLIB::ZLIB
                 APPEND PROPERTY INTERFACE_INCLUDE_DIRECTORIES
                 $<$<CONFIG:Release>:${zlib_INCLUDE_DIRS_RELEASE}>)
    # Necessary to find LINK shared libraries in Linux
    set_property(TARGET ZLIB::ZLIB
                 APPEND PROPERTY INTERFACE_LINK_DIRECTORIES
                 $<$<CONFIG:Release>:${zlib_LIB_DIRS_RELEASE}>)
    set_property(TARGET ZLIB::ZLIB
                 APPEND PROPERTY INTERFACE_COMPILE_DEFINITIONS
                 $<$<CONFIG:Release>:${zlib_COMPILE_DEFINITIONS_RELEASE}>)
    set_property(TARGET ZLIB::ZLIB
                 APPEND PROPERTY INTERFACE_COMPILE_OPTIONS
                 $<$<CONFIG:Release>:${zlib_COMPILE_OPTIONS_RELEASE}>)

########## For the modules (FindXXX)
set(zlib_LIBRARIES_RELEASE ZLIB::ZLIB)
```

</details>

Reply:
Actually, it does now set the CMAKE_BUILD_TYPE. I merged a PR for that right before I reviewed this :)

Reply:
Ok, so, do I change it too? The only important question is if this works with the multiple build configurations project files for `MSBuild`

Reply:
I honestly have no idea what the interaction between the CMake module and the multi configuration output of the vs backend is.

Reply:
But I do think it makes sense to do something more like `lambda x: '1' if buildtype_map[x] == 'Debug' else '0'`

Reply:
Yeah, but what about all the exotic build types (release with debug info, minsize...)?

Reply:
Either that or just tie it to the `get_option('debug')` option? Not sure what the right thing to do here is honestly, and the person who really understood this code isn't really around anymore

Reply:
I just rebased onto the laetst `master` from today and I still have `trace.cache['CMAKE_BUILD_TYPE']` = `CMakeCacheEntry(value=[''], type='STRING')` here

Normally, in `CMake`, these match (case-insensitively) on the configuration name. `conan` does the same - you can name your configuration (buildtype) whatever you like, and `conan` will produce config files that use this configuration name. You are simply highly advised to use `Release` and `Debug` unless you have a good reason to change.

Reply:
Yeah, it is possible, but I've never actually seen it happen. I'm happy to punt on those kinds of corners until we actually need to deal with them